### ISSUE TITLE
Use bundler

### ISSUE THREAD

**Issue by DavideCarvalho** [openzipkin/zipkin-js#270]
Maybe zipkin libs should use a bundler like [rollup](https://rollupjs.org/guide/en). This will make libraries lightier because of tree shaking and dead code elimination. 
Front end applications should have a big improvement because of this.

Relates to #195 (comment)

**Comment by jcchavezs** [comment]
Thanks @DavideCarvalho would you open a PR for drafting this? Any idea whether this would impact our releasing process?

**Comment by jcchavezs** [comment]
Relates to https://github.com/openzipkin/zipkin-js/issues/195#issuecomment-431528258

**Comment by DavideCarvalho** [comment]
Okay, i'll do it once require.call PR is over

**Comment by DavideCarvalho** [comment]
I'm make this PR ASAP.
We are using zipkin js at work and trying to load it with SystemJS. The way it is bundled today it's really making things really hard for us.
I cloned the repo and made a rollup configuration for zipkin to generate .umd, cjs e esm files (the main is cjs).
If you prefer, we could still build the way it is built now a days and just add the rollup build as well.

**Comment by jcchavezs** [comment]
Awesome @DavideCarvalho !

**Comment by DavideCarvalho** [comment]
Maybe create a git-hook with husky to generate the bundle before commiting should be a good strategy for now.

**Comment by DavideCarvalho** [comment]
Husky really should be a thing. Building before commiting to ensure nothing is break is a good idea.

**Comment by adriancole** [comment]
Davi we are very grateful for your help. Rock on!

**Comment by DavideCarvalho** [comment]
@jcchavezs I'm not saying to drop version 4 compatibility, but node 4.x doesn't let us build.
Is there any way we can build with node>4.x and test it with 4.x?

**Comment by adriancole** [comment]
I think we should use our devops group when asking questions like this else
things get pinned to the creativity and accessibility of individuals ;)

**Comment by DavideCarvalho** [comment]
Now that node4 has dropped, we can try to bundle the library again.
I guess we can try to bundle ts and js files together, (while the codebase is not entirely ts) see if it’s possible.

**Comment by jcchavezs** [comment]
[Comment](https://github.com/openzipkin/zipkin-js/issues/16#issuecomment-441215893) from @sonybinhle:

> I am for rollup also.

**Comment by jcchavezs** [comment]
Would be nice to have some input from @openzipkin/zipkin-js-champions and @ghermeto

**Comment by DavideCarvalho** [comment]
This shouldn't be a difficult task. Rollup has a nice example that we can basically copy and paste. I don't know how it works with .ts and .js files together, tho, but I don't think it should be a problem for Rollup.

**Comment by sonybinhle** [comment]
@jcchavezs Could I take it? if it is still not assigned to anyone. 💃

**Comment by DavideCarvalho** [comment]
@sonybinhle go for it! I guess no one is doing it

**Comment by jcchavezs** [comment]
Yes please!

José Carlos Chávez


Den ons. 5. des. 2018 kl. 13:03 skrev Davi de Carvalho <
notifications@github.com>:

**Comment by webmutation** [comment]
https://www.npmtrends.com/rollup-vs-webpack

Rollup use is kind of declining. Just wanted to leave this here since browserify was picked last time was also declining. In any case they are very similar.

**Comment by DavideCarvalho** [comment]
@webmutatation this is interesting, but nowadays as far as I know, rollup it the only one who can output some types of modules, like esm and mjs. Do you know any alternatives for these?

**Comment by jcchavezs** [comment]
So after #270 is merged, what is missing here?

**Comment by sonybinhle** [comment]
@jcchavezs I currently use Rollup for Packages/zipkin. The rest are still using the old way of bundling. Should I start adding Rollup for the rest packages?

**Comment by DavideCarvalho** [comment]
For now, I don't see why not keep using Rollup for the other packages. If we see that rollup doesn't work with other package, we can discuss

**Comment by jcchavezs** [comment]
Good. @sonybinhle, could you work on the other packages?

José Carlos Chávez


Den ons. 12. des. 2018 kl. 11:58 skrev Davi de Carvalho <
notifications@github.com>:

**Comment by sonybinhle** [comment]
@DavideCarvalho @jcchavezs Thank you very much for your feedback. I will work on it.

**Comment by jcchavezs** [comment]
Awesome!

José Carlos Chávez


Den ons. 12. des. 2018 kl. 12:35 skrev Binh Le <notifications@github.com>:

**Comment by webmutation** [comment]
Hi David, Webpack supports thru plugins probably all existing module formats. 
I think mjs is out of the box and for esm you have a plugin
https://www.npmjs.com/package/@purtuga/esm-webpack-plugin

But I actually prefer Rollup it is cleaner output (webpack is a bit messy), just not as popular and a bit in decline and also makes my use case Angular CLI app a bit harder to integrate, but i have forked so my particular needs do not block anything here :) I actually prefer Parcel because of zero config, but that is another story. 

Yes, please use only one for all packages, in this case Rollup.

@jcchavezs when will there be a new release 0.16 or will you push for 1.0.0 ? I have some issues currently with one of the libs used in zipkin, I think it is documented in #155 so I will not open a new one.

Big thanks to all you guys for your efforts, this project is really gaining some traction :1st_place_medal:

**Comment by ghermeto** [comment]
can anyone please explain the need to use `rollup` (or any bundler for that matter) for packages that will work exclusively for Node.js? 

I would understand using rollup for `zipkin-transport-http` and others that can be used on a browser, but most of the other packages would be better off without it IMO. 

It is not like you can compress the code, as it is a bad practice for node because makes installed packages inspection harder and will drive companies that rely on such inspections away (and to build their own versions of zipkin client).

**Comment by jcchavezs** [comment]
This sound like I really valid concern, should we consider on only minify
those libraries that can be used in frontend and leave the rest out? Any
opinion @openzipkin/zipkin-js-champions

Den ons. 12. des. 2018, 18:10 skrev Guilherme Hermeto <
notifications@github.com:

**Comment by jcchavezs** [comment]
Based on @ghermeto's concerns. Could we discuss which libraries should support bundler?

- [ ] cujojs/rest (zipkin-instrumentation-cujojs-rest)
- [ ] express (zipkin-instrumentation-express)
- [ ] fetch (zipkin-instrumentation-fetch)
- [ ] hapi (zipkin-instrumentation-hapi)
- [ ] memcached (zipkin-instrumentation-memcached)
- [ ] redis (zipkin-instrumentation-redis)
- [ ] restify (zipkin-instrumentation-restify)
- [ ] postgres (zipkin-instrumentation-postgres)
- [ ] request (zipkin-instrumentation-request)
- [ ] connect (zipkin-instrumentation-connect)
- [ ] superagent (zipkin-instrumentation-superagent)
- [ ] grpc-client (zipkin-instrumentation-grpc-client)
- [ ] zipkin-transport-http
- [ ] zipkin-transport-kafka
- [ ] zipkin-transport-scribe
- [ ] zipkin-encoder-thrift
- [ ] zipkin-context-cls

**Comment by sonybinhle** [comment]
@jcchavezs 
IMO there are advantages and disadvantages of using bundler for all packages.

By using bundler, we can export final bundle to different formats, uniform way of bundling code, plugin supports. It will be at the expense of simplicity, readability output(a little bit readability when babel transpile file by file instead of combining all of the files to one).

I prefer using the same bundler for all packages since the trade-off is not significant. We can have a configuration to say that should we build a UMD format or not for packages used in server only.

**Comment by ghermeto** [comment]
The consistency achieved is not a reason strong enough for the price you pay with loss of simplicity, readability, debuggability and security inspections. 

It is desirable for client side javascript, but definitely not for packages that will only be used by Node.js.
